### PR TITLE
Refactor ModifyQueueCapacityAction to follow builder pattern

### PR DESCRIPTION
**Pull request by rguo-aws**
*Issue #:*

*Description of changes:*
Refactor ModifyQueueCapacityAction to follow builder pattern. the current constructor in Modify**Action does not allow decider to make changes to some parameters in Action object. So I am creating this PR to add a Builder into the Modify**Action class to expose more parameters for decider to tune

*Tests:*
n/a

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java [hunk 1]
```diff
@@ -127,4 +112,103 @@ public int getDesiredCapacity() {
...
+      this.stepSize = DEFAULT_STEP_SIZE;
+      this.isIncrease = DEFAULT_IS_INCREASE;
+      this.desiredCapacity = null;
+      this.currentCapacity =
+          NodeConfigCacheReaderUtil.readQueueCapacity(appContext.getNodeConfigCache(), esNode, threadPool);
+      setDefaultBounds(threadPool);
+    }
+
+    private void setDefaultBounds(ResourceEnum threadPool) {
+      // TODO: Move configuration values to rca.conf
```
[the remark this answers is not in the document]
Comment (vigyasharma) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:147]: Let's create a `QueueConfig` class that can extract all queue related configs from `RcaConf` and apply defaults/overrides. We don't want to clutter `rca.conf` with default values, but also don't want to scatter queue config across multiple places. The queue builder will then just accept a `QueueConfig` object and use bounds from it.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java [hunk 2]
```diff
@@ -127,4 +112,103 @@ public int getDesiredCapacity() {
...
+    public Builder setLowerBound(int lowerBound) {
+      this.lowerBound = lowerBound;
+      return this;
+    }
+
+    public ModifyQueueCapacityAction build() {
+      //fail to read capacity from node config cache
+      // return null action
+      if (currentCapacity == null) {
+        return null;
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java]: done. Action builder will return a non-actionable default action if it fails to read current capacity from node config cache

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java [hunk 3]
```diff
@@ -127,4 +112,103 @@ public int getDesiredCapacity() {
...
+        case SEARCH_THREADPOOL:
+          this.lowerBound = 1000;
+          this.upperBound = 3000;
+          break;
+        default:
+          assert false : "unrecognized threadpool type: " + threadPool.name();
+      }
+    }
+
+    public Builder setCoolOffPeriod(long coolOffPeriodInMillis) {
```
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java]: done renamed those method to remove the set prefix

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:156]: As per our discussion, let's track this in a separate issue. Action build should keep a reference of RcaConf class object and read the corresponding action configs(or default configs) from the conf object. The RcaConf object can be either passed in from decider or we can add the RcaConf into AppContext and share it within the process.

[the remark this answers is not in the document]
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:156]: issue created #367

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java [hunk 4]
```diff
@@ -127,4 +118,108 @@ public int getDesiredCapacity() {
...
+          assert false : "unrecognized threadpool type: " + threadPool.name();
+      }
+    }
+
+    public Builder coolOffPeriod(long coolOffPeriodInMillis) {
+      this.coolOffPeriodInMillis = coolOffPeriodInMillis;
+      return this;
+    }
+
+    public Builder increase(boolean isInrease) {
```
Comment (sruti1312) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:176]: isIncrease*

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java [hunk 5]
```diff
@@ -23,34 +23,43 @@
...
-      boolean increase, AppContext appContext) {
+  private static final Logger LOG = LogManager.getLogger(ModifyQueueCapacityAction.class);
+  public static final String NAME = "ModifyQueueCapacity";
+
+  private final ResourceEnum threadPool;
+  private final NodeKey esNode;
+  private final int desiredCapacity;
+  private final int currentCapacity;
+  private final int lowerBound;
+  private final int upperBound;
```
Comment (sruti1312) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:43]: Why do we want lower and upper bound here? Builder takes these values in consideration for calculating the desired value

Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:43]: one use case is the decider might use this upper/lower bound to calculate the proper step size for each action. We don't use it at this moment but let's say JVM decider will bucketize the current capacity between the [lower, upper] range and based the bucket this resource is allocated, the decider might choose different step size to tune the resource

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java (hunk 4 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:176]: thanks, will fix this

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java (hunk 5 above)
Comment (rguo-aws) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/actions/ModifyQueueCapacityAction.java:43]: we can remove them later if decider is end up not using it

**Pull request merged**